### ISSUE TITLE
forecasting next couple time steps

### ISSUE THREAD

**Issue by TiaHao** [icc2115/Neural-GC#16]
Dear developer,

I am currently using the NGC framework to find the GC relationship and doing next 1 time step forecasting, but now I would like to extend to next 3-6 time steps forecasting, would this be possible to modify these model structures for this purpose? thank you!

**Comment by iancovert** [comment]
Yes, this should be possible. The things you'll need to change include 1) the model's `__init__` function to ensure it outputs predictions for multiple time steps, and 2) how the loss is calculated. Otherwise, the same idea of sparsifying each network via its input layer seems valid. 

For the model setup, you can see the cMLP source code [here](https://github.com/iancovert/Neural-GC/blob/master/models/cmlp.py#L8) for example. The forward function applies a series of fully-connected layers (implemented via Conv1d layers to do so in parallel for all time steps), including a final layer that outputs a single value per time step. To change this, you would modify the part of [this line](https://github.com/iancovert/Neural-GC/blob/master/models/cmlp.py#L17) with `hidden[1:] + [1]` to `hidden[1:] + [num_steps]`. You'll also want to make sure the cMLP class, which wraps the MLP class, concatenates the outputs properly [here](https://github.com/iancovert/Neural-GC/blob/master/models/cmlp.py#L62).

For the loss calculation, you'll want to modify the part of the training code that handles the forecasting MSE. For the cMLP's GISTA training, it's handled [here](https://github.com/iancovert/Neural-GC/blob/master/models/cmlp.py#L241). The first argument to `loss_fn` is the predictions, which should be handled by your changes above. The second argument is the targets, and you'll want to rearrange the time series differently than `X[:, lag:, i:i+1]` to have targets for the multiple forecasts.

Let me know if this makes sense.

**Comment by TiaHao** [comment]
thanks for your timely reply! Yes it makes sense a lot. I am also modifying it as you suggested. However, I notice that the computational time of multi-step forecasting of cMLP dramatically increases, so I am thinking if this model structure is not very suitable for multi-step forecasting purpose, especially if I only care about identifying GC relationship and forecasting for 1 feature.

**Comment by iancovert** [comment]
That's interesting, let me think about whether it makes sense for the forecasting time to dramatically increase...I suppose that the computation for each prediction has more FLOPs in the second matmul (implemented as a convolution), so this could perhaps as much as double the computation time. Is the slowdown you're seeing consistent with that, or even worse?

If you're only interested in identifying GC relationships when forecasting a single variable, you're right that you could simplify the problem. Rather than fitting an entire cMLP with one network per variable, you could fit a single MLP for the variable you want to forecast, and train it with the same regularization approach.
